### DISCUSSION TITLE
[Proposal] Strict switch requiring to handle all possible options (if applicable)

### DISCUSSION THREAD

**Discussion by wojciechsura** [dotnet/csharplang#6164] [Language Ideas]
I'd like to propose a strict switch version, which requires all enumeration values to be handled. The problem, which this construct might solve looks like following:

```csharp
public enum SomeEnum
{
    One,
    Two,
    Three
}

// (...)

public void Test()
{
    SomeEnum val = SomeEnum.One;

    switch (val)
    {
         case SomeEnum.One:
             break;
         case SomeEnum.Two:
             break;
         case SomeEnum.Three:
             break;
    }
}
```

If one added an additional value to SomeEnum, the Test method might behave erratically, because it doesn't handle the newly added item.

```csharp
public enum SomeEnum
{
    One,
    Two,
    Three,
    Four
}

// (...)

public void Test()
{
    SomeEnum val = SomeEnum.Four;

    // Most likely undesired behavior
    switch (val)
    {
         case SomeEnum.One:
             break;
         case SomeEnum.Two:
             break;
         case SomeEnum.Three:
             break;
    }
}
```

This can be solved by adding a default case with exception:

```csharp
public enum SomeEnum
{
    One,
    Two,
    Three,
    Four
}

// (...)

public void Test()
{
    SomeEnum val = SomeEnum.Four;

    // Most likely undesired behavior
    switch (val)
    {
         case SomeEnum.One:
             break;
         case SomeEnum.Two:
             break;
         case SomeEnum.Three:
             break;
         default:
             throw new InvalidEnumArgumentException("Unsupported SomeEnum value!");
    }
}
```

This however has a drawback, because the problem *might* be found in the runtime if the method in question is actually executed.

I propose a now notion (to be discussed), which requires developer to handle all possible values, or add a default case.

```csharp
public enum SomeEnum
{
    One,
    Two,
    Three,
    Four
}

// (...)

public void Test()
{
    SomeEnum val = SomeEnum.Four;

    // Compilation error: not all cases were handled
    switch strict (val)
    {
         case SomeEnum.One:
             break;
         case SomeEnum.Two:
             break;
         case SomeEnum.Three:
             break;
    }

    // OK - all cases were handled
    switch strict (val)
    {
         case SomeEnum.One:
             break;
         case SomeEnum.Two:
             break;
         case SomeEnum.Three:
             break;
         case SomeEnum.Four:
             break;
    }

    // OK - default was provided
    switch strict (val)
    {
         case SomeEnum.One:
             break;
         case SomeEnum.Two:
             break;
         case SomeEnum.Three:
             break;
         default:
             break;
    }

}
```

In case of other enumerable types (like int, byte etc.), the behavior would be the same - requirement to handle all values. Ranges might be helpful in such cases:

```
    switch strict (byteVal)
    {
        case [0..31]:
            return "Control ASCII";
        case [32..127]:
            return "Readable ASCII";
        case [128..255]:
            return "Variable ASCII";
    }
```

This is non-breaking change, because a completely new notion would be introduced. The "regular" `switch` would work exactly the same way as it did.

I'm interested, if you see this functionality (which probably shouldn't be too hard to implement) is of an interest to you.

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/3179

**Reply by wojciechsura** [reply]
@ufcpp It may seem similar in the beginning, but I believe that my proposal causes *far* less problems for all: implementation, handling existing standard library code, existing libraries, codebases, cast handling and so on.

Namely:

1. Closed enum forces you to decide at the point of enum *declaration*, if it is required to have all cases handled, whereas strict switch allows you to do that at the point where you are actually switching through the enum.
2. Strict switch allows you to easily mix switches and strict switches. Closed enum forces you to use strict switches everywhere. This may lead to people writing two version of enums - one closed and one open, which hinders code quality.
3. Introduction of strict switch doesn't require to change standard library code. In case of closed enums, should standard library enums be converted to closed or not? If so, it will break existing code. If not, standard library won't get any benefit of having this construct. In case of strict switch, enums may be left as they are, and library code can (but doesn't have to) be changed to use strict switches without literally *any* impact on existing code.
4. Strict switch doesn't introduce any problems with casting, whereas closed enums - as the thread suggests - does
5. Strict switch statement follows existing language's feature of switch expression, which is "softly strict" as in yielding a warning, when all cases are not handled.
6. Strict switch allows strictly switching through anything enumerable, namely enums, integer types, char, bool and all of those with nullable variant (as in eg. bool?). Closed enums doesn't give that level of flexibility.

I would gladly hear out any arguments for closed enums instead of strict switches, because honestly I see none...

**Comment by wojciechsura** [comment]
An interesting thing: Kotlin has this feature by default for its equivalent of switch statement/expression: https://kotlinlang.org/docs/control-flow.html#when-expression

**Reply by HaloFour** [reply]
I don't see any real difference here between Kotlin `when` or C# `switch`.  The language enforces that you cover the known cases, and if your check is not exhaustive you are required/suggested to add a `default`/`else` case.  The real difference lies in how `enum` works between Kotlin/Java and C#.  With the former it's practically impossible to have an invalid value, other than `null` (and maybe some really abusing reflection).  With the latter it's relatively easy and can be caused by a cast or bitwise operation.

IMO I still disagree with the language designers that `switch` needs to handle the undefined cases here.  Sure, it's possible, but it's a lot less likely and seemingly indicative of some other kind of problem.  For `switch` expressions the compiler will emit a `default` case anyway, which will thrown.  My preference, rather than trying to define new flavors of `enum` or `switch`, would be to reconsider this warning in a future version as the compiler implements support for more closed sets like discriminated unions.

I just realized that the C# compiler emits two different kinds of warnings, one if the `switch` covers all enum cases and one where it doesn't.  It seemingly would be possible to suppress one of these warnings and to achieve the behavior that all enums within the project or specific parts of the project are treated as a closed set:

```cs
E e = E.A;

// no warning
int result1 = e switch {
    E.A => 1,
    E.B => 2,
    _ => 3
};

// warning: CS8524 - '(E)2' is not covered.
int result2 = e switch {
    E.A => 1,
    E.B => 2
};

// warning: CS8509 - 'E.B' is not covered.
int result3 = e switch {
    E.A => 1
};

#pragma warning disable CS8524
// no warning
int result4 = e switch {
    E.A => 1,
    E.B => 2
};


enum E { A, B }
```
